[WIP] Linux/BSD: Add status indicator support through DBus

deralmas opened the pull request:
This PR adds support for status indicators in the `FreeDesktopPortalDesktop` class.

While it's technically not part of the portal API, splitting it would've been too inconvenient for now. I'd like to eventually put most of the abstractions/tools into a separate class (a bit like a homebrew binding) or at worst rename it to something more generic like `FreeDesktopDBus`.

## Changes to event handling

This PR also fleshes further the "core" of the DBus handling logic, as the StatusNotifierItem API is very object-oriented[^1].

Most importantly, I changed the monitor loop from a polling loop to a `poll(2)`-based one (no idea how it's properly named), by tracking the various `DBusWatch`es that libdbus throws at us. 

Note that technically we should also track `DBusTimeout`s ~but I did not investigate on what they're for yet. Stuff works for now so :shrug:~

Edit: they're for timeouts duh xD

I also removed most of the "simple" event logic (`dbus_connection_pop_message`) to go through the "builtin" event handlers, as otherwise we would've had to rebuild the event handling logic from scratch.

Finally, I added a bunch of utility methods to populate common "variant" (the DBus kind) based messages . I wonder if it would be worth it to make a fancier `Variant`-based API/binding of sorts as the feature list grows (I want to plumb in the whole portal API eventually) or if that's a bad idea.

## Final remarks

Despite it being technically feature-complete (although I didn't test the tooltip because swaybar does not support it), I marked this PR as WIP because the code is still a bit rough, it needs some more testing and I only plumbed it into the Wayland backend (should be trivial to do the same on X11 though).

Most importantly, I'd like to discuss the changes I did to the DBus event loop and whether I should split them up into a separate PR.

CC @bruvzg

## TODO

 - [x] Fix issue where the icons don't get reassigned when the bar (e.g. swaybar) gets restarted
 - [ ] Implement DBusTimeout support (the docs say what we should do but not why, doesn't sound too hard)
 - [ ] Investigate in some prettier way of writing OO interactions (custom bindings?)

[^1]: https://www.freedesktop.org/wiki/Specifications/StatusNotifierItem/

deralmas committed:
[WIP] Linux/BSD: Add status indicator support through DBus

AThousandShips commented on platform/linuxbsd/freedesktop_portal_desktop.h:
```diff
@@ -95,6 +144,12 @@ class FreeDesktopPortalDesktop : public Object {
+
+	bool indicator_create(DisplayServer::IndicatorID p_id, const Ref<Image> &p_icon);
+	Error indicator_set_icon(DisplayServer::IndicatorID p_id, const Ref<Image> &p_icon);
+	void indicator_set_tooltip(DisplayServer::IndicatorID p_id, String p_tooltip);
```
```suggestion
	void indicator_set_tooltip(DisplayServer::IndicatorID p_id, const String &p_tooltip);
```

AThousandShips commented on platform/linuxbsd/freedesktop_portal_desktop.cpp:
```diff
@@ -617,6 +749,262 @@ void FreeDesktopPortalDesktop::_system_theme_changed_callback() {
+	MutexLock mutex_lock(file_dialog_mutex);
+
+	ERR_FAIL_COND_V(!indicators.has(p_id), ERR_UNCONFIGURED);
+	ERR_FAIL_COND_V(p_icon->get_size() == Size2i(), ERR_INVALID_PARAMETER);
```
This won't fire if the size is `{ 0, 1 }` or `{ 1, 0 }` etc., only if both axes are zero

bruvzg commented on platform/linuxbsd/freedesktop_portal_desktop.cpp:
```diff
@@ -617,6 +749,263 @@ void FreeDesktopPortalDesktop::_system_theme_changed_callback() {
+
+	item.icon_data.clear();
+
+	for (int y = 0; y < image->get_width(); ++y) {
+		for (int x = 0; x < image->get_width(); ++x) {
+			Color pixel = image->get_pixel(x, y);
+
+			item.icon_data.append(pixel.get_a8());
+			item.icon_data.append(pixel.get_r8());
+			item.icon_data.append(pixel.get_g8());
+			item.icon_data.append(pixel.get_b8());
+		}
+	}
```
`get_pixel` have a bunch of unnecessary checks + color conversion to float `Color` and is pretty slow, so in the similar cases we usually use:

```
image->convert(Image::FORMAT_RGBA8);
const uint8_t *r = image->get_data().ptr();
```

and accessing RGBA8 data from `r` pointer directly.

bruvzg replied:
Names are most likely from GL - https://registry.khronos.org/OpenGL-Refpages/gl4/html/glTexImage2D.xhtml

bruvzg commented on platform/linuxbsd/freedesktop_portal_desktop.cpp:
```diff
@@ -617,6 +749,263 @@ void FreeDesktopPortalDesktop::_system_theme_changed_callback() {
+				// The documentation says that this is pretty much an app id.
+				success = _dbus_messsage_iter_append_basic_variant(&reply_iter, DBUS_TYPE_STRING, "org.godotengine.Godot");
+			} else if (strcmp(property, "Title") == 0) {
+				// TODO: No idea what would be a good value.
+				// Same thing as above, although it's meant to be a little bit more verbose.
+				success = _dbus_messsage_iter_append_basic_variant(&reply_iter, DBUS_TYPE_STRING, "Godot game engine");
```
I guess it should be something like this:
https://github.com/godotengine/godot/blob/29b3d9e9e538f0aa8effc8ad8bf19a2915292a89/platform/linuxbsd/tts_linux.cpp#L59-L65

ID can be generated from the same project settings (it's done in multiple places for other platforms), but I feel like we might want it as a dedicated project setting, since unlike title, IDs are usually limited to basic Latin, numbers and a few separator characters.

bruvzg commented on platform/linuxbsd/freedesktop_portal_desktop.cpp:
```diff
@@ -617,6 +749,263 @@ void FreeDesktopPortalDesktop::_system_theme_changed_callback() {
+	ERR_FAIL_COND_V(!portal->indicators.has(id), DBUS_HANDLER_RESULT_HANDLED);
+	StatusNotifierItem &item = portal->indicators[id];
+
+	if (dbus_message_is_method_call(message, BUS_INTERFACE_STATUS_NOTIFIER_ITEM, "Activate")) {
```
Probably should also handle `SecondaryActivate` for other mouse button clicks (not sure if there is a way to determine exact mouse button, but it's probably OK to call cb with MIDDLE button for it).

bruvzg replied:
I think activation callback on macOS/Windows have position and mouse button arguments, this events should have position as well and we can set mouse button to RIGHT for primary and MIDDLE (or LEFT) for secondary (mouse button is not precise on Windows as well, but primary activation pretend to be right click).

bruvzg commented on platform/linuxbsd/freedesktop_portal_desktop.cpp:
```diff
@@ -617,6 +749,263 @@ void FreeDesktopPortalDesktop::_system_theme_changed_callback() {
+}
+
+Error FreeDesktopPortalDesktop::indicator_set_icon(DisplayServer::IndicatorID p_id, const Ref<Image> &p_icon) {
+	MutexLock mutex_lock(file_dialog_mutex);
```
Mutex probably should be renamed to something generic, originally it was intended to guard `file_dialogs` array writes only.

bruvzg commented on platform/linuxbsd/freedesktop_portal_desktop.h:
```diff
@@ -72,11 +75,57 @@ class FreeDesktopPortalDesktop : public Object {
+
+	struct StatusNotifierItem {
+		DisplayServer::IndicatorID id = DisplayServer::INVALID_INDICATOR_ID;
+		DBusConnection *connection = nullptr;
```
I guess the same logic as in the https://github.com/godotengine/godot/pull/89935/ applies here, and there's no point to have separate connection for the item, since it will be the same for all of them anyway.

deralmas committed:
Handle bar reloads

deralmas committed:
Remove the name_assigned thing

It does not make sense.

This logic was originally meant for tracking whether we're registered or
not but we can't really do that and it's not worth it.

Also it broke the build xD

deralmas committed:
Make a variable name more correct

Small nit but this spec is already confusing as it is

deralmas committed:
Hopefully fix weird vaarg issue

The thing required a `char**` but for some reason worked on GCC. Weird.

deralmas committed:
Fix arg pointer mess for good

Ugh the last commit didn't even compile. You know what, I'll stop trying
to be clever. If this wants a `const char **`, then let's just go to the
simple (but verbose) route.

deralmas committed:
Fix various things after rebase

deralmas committed:
Don't unref handled messages

I chased this thing for ages. It looks like dbus unrefs the handled
message by itelf, because if I do it myself everything explodes. This is
completely undocumented, supposing it's even what's happening. WTF
